src/ledger/GeneralizedLedgerEntry.h

```diff
@@ -0,0 +1,90 @@
+// Copyright 2020 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#pragma once
+
+#include "xdr/Stellar-ledger-entries.h"
+#include "xdr/Stellar-ledger.h"
+
+namespace stellar
+{
+
+enum class GeneralizedLedgerEntryType
+{
+    LEDGER_ENTRY
+};
+
+class GeneralizedLedgerKey
+{
+  private:
+    GeneralizedLedgerEntryType mType;
+    union {
+        LedgerKey mLedgerKey;
+    };
+
+    void assign(GeneralizedLedgerKey const& glk);
+    void construct();
+    void destruct();
+
+    void checkDiscriminant(GeneralizedLedgerEntryType expected) const;
+
+  public:
+    GeneralizedLedgerKey(LedgerKey const& lk);
```

Comment:
shouldn't all this stuff be `explicit`?

Reply:
Well, it depends. I considered all approaches. In order to get compatibility between "generalized" and "non-generalized" we need to do one of the following:
1. do not use explicit here 
2. add overloads everywhere that we want compatibility
3. change all the existing operations to explicitly convert to "generalized"

Path 3 is a painful option. Path 2 is a possible route but seems like a lot of work compared to path 1. We can always change this retroactively. Choosing path 1 means we aren't creating extra work by doing path 2 or path 3 in the future.

It's important to note that all the other constructors (eg. for CAP-33 sponsorship keys) will be explicit because there is no existing code that depends on it.